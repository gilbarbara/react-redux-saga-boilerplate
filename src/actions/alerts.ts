import { ReactNode } from 'react';
import { uuid } from '@gilbarbara/helpers';
import { createAction } from '@reduxjs/toolkit';

import { actionPayload } from '~/modules/helpers';

import { ActionTypes } from '~/literals';

import { ShowAlertOptions } from '~/types';

export const hideAlert = createAction<string>(ActionTypes.ALERTS_HIDE);

export const showAlert = createAction(
  ActionTypes.ALERTS_SHOW,
  (message: ReactNode, options: ShowAlertOptions) => {
    const timeout = options.type === 'error' ? 0 : 5;

    return actionPayload({
      id: options.id || uuid(),
      icon: options.icon || 'info-o',
      message,
      position: options.position || 'bottom-right',
      type: options.type || 'neutral',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    });
  },
);
