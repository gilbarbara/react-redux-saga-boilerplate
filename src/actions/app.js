// @flow
/**
 * @module Actions/App
 * @desc App Actions
 */

import uuid from 'uuid/v4';
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export { goBack, go, push, replace } from 'modules/history';

export const { hideAlert, showAlert, switchMenu } = createActions({
  [ActionTypes.SWITCH_MENU]: (query: string) => ({ query }),
  [ActionTypes.HIDE_ALERT]: (id: string) => ({ id }),
  [ActionTypes.SHOW_ALERT]: (message: string, options: Object) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || uuid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
});
