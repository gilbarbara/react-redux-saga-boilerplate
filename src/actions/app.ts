/**
 * @module Actions/App
 * @desc App Actions
 */
import React from 'react';
import { nanoid } from 'nanoid';

import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { ShowAlertOptions } from 'types';

export { go, goBack, push, replace } from 'modules/history';

export const hideAlert = createAction(ActionTypes.HIDE_ALERT, (id: string) => ({ id }));

export const showAlert = createAction(
  ActionTypes.SHOW_ALERT,
  (message: React.ReactNode, options: ShowAlertOptions) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || nanoid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
);
