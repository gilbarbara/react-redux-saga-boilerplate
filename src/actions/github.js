// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { githubGetRepos: getRepos } = createActions({
  [ActionTypes.GITHUB_GET_REPOS]: (query: string) => ({ query }),
});
