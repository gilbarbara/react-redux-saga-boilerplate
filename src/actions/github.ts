/**
 * @module Actions/User
 * @desc User Actions
 */
import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

export const getRepos = createAction(ActionTypes.GITHUB_GET_REPOS_REQUEST, (query: string) => ({
  query,
}));
