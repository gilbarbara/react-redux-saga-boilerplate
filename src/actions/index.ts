export * from './app';
export * from './github';
export * from './user';
