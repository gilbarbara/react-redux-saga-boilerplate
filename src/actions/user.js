// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { userLogin: login, userLogout: logOut } = createActions({
  [ActionTypes.USER_LOGIN]: () => ({}),
  [ActionTypes.USER_LOGOUT]: () => ({}),
});
