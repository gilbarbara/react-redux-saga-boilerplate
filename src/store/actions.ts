export { alertHide, alertShow } from './slices/alerts';
export { setAppOptions } from './slices/app';
export { getRepos, getReposFailure, getReposSuccess } from './slices/github';
export { login, loginSuccess, logOut, logOutSuccess } from './slices/user';
