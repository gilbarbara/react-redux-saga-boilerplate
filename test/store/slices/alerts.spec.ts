import { emptyAction } from 'test-utils';

import { alertHide, alertShow } from '~/actions';

import reducer from '~/store/slices/alerts';

describe('slices/alerts', () => {
  describe('actions', () => {
    describe(`${alertShow.type}`, () => {
      it('should return an action with variant `error`', () => {
        expect(alertShow('Alright!', { id: 'test', type: 'error' })).toMatchSnapshot();
      });

      it('should return an action with variant `success`', () => {
        expect(
          alertShow('Alright!', { id: 'test', type: 'success', timeout: 10 }),
        ).toMatchSnapshot();
      });
    });

    describe(`${alertHide.type}`, () => {
      it('should return an action', () => {
        expect(alertHide('test')).toMatchSnapshot();
      });
    });
  });

  describe('reducer', () => {
    let alerts = reducer(undefined, emptyAction);

    it('should return the initial state', () => {
      expect(reducer(alerts, emptyAction)).toMatchSnapshot();
    });

    it(`should handle "${alertShow.type}"`, () => {
      alerts = reducer(alerts, alertShow('HELLO', { id: 'test', type: 'success' }));
      expect(alerts).toMatchSnapshot();
    });

    it(`should handle "${alertHide.type}"`, () => {
      alerts = reducer(alerts, alertHide('test'));
      expect(alerts).toMatchSnapshot();
    });
  });
});
