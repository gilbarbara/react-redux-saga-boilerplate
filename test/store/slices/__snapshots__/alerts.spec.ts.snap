// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`slices/alerts > actions > alerts/alertHide > should return an action 1`] = `
{
  "payload": "test",
  "type": "alerts/alertHide",
}
`;

exports[`slices/alerts > actions > alerts/alertShow > should return an action with variant \`error\` 1`] = `
{
  "meta": undefined,
  "payload": {
    "icon": "info-o",
    "id": "test",
    "message": "Alright!",
    "position": "bottom-right",
    "timeout": 0,
    "type": "error",
  },
  "type": "alerts/alertShow",
}
`;

exports[`slices/alerts > actions > alerts/alertShow > should return an action with variant \`success\` 1`] = `
{
  "meta": undefined,
  "payload": {
    "icon": "info-o",
    "id": "test",
    "message": "Alright!",
    "position": "bottom-right",
    "timeout": 10,
    "type": "success",
  },
  "type": "alerts/alertShow",
}
`;

exports[`slices/alerts > reducer > should handle "alerts/alertHide" 1`] = `
{
  "data": [],
}
`;

exports[`slices/alerts > reducer > should handle "alerts/alertShow" 1`] = `
{
  "data": [
    {
      "icon": "info-o",
      "id": "test",
      "message": "HELLO",
      "position": "bottom-right",
      "timeout": 5,
      "type": "success",
    },
  ],
}
`;

exports[`slices/alerts > reducer > should return the initial state 1`] = `
{
  "data": [],
}
`;
